migrate.py

```diff
@@ -0,0 +1,217 @@
+
+        ds_mountpoint = _eval(f"zfs get -Hp -o value mountpoint '{dataset_name}'").strip()
+        snapshot_path = "/".join([ds_mountpoint, ZFS_SNAPSHOTDIR, snapshot_name])
+
+        snapshot_time = _eval(f"zfs get -Hp -o value creation '{full_snapshot_name}'")
+        snapshot_time_readble = str(datetime.datetime.fromtimestamp(int(snapshot_time)))
+
+        # Use proot to "mount" coorect path. See https://github.com/restic/restic/issues/2092
+        proot_command = f"proot -b '{snapshot_path}':'{path_in_restic_repo}'"
+        restic_backup_args = ["--ignore-ctime", "--time", snapshot_time_readble, "--tag", f"{SNAPSHOT_TAG}{snapshot_name}"]
+        if parent_restic_snapshot is not None:
+            restic_backup_args += ["--parent", parent_restic_snapshot]
+        restic_backup_args.append(path_in_restic_repo)
+        restic_command = self._restic_cmd(restic_repo, "backup", restic_backup_args)
+        print(f"Starting backup of {dataset_name}@{snapshot_name} into {restic_repo} under {path_in_restic_repo}")
+        _run(f"{proot_command} {restic_command}")
+
+    def backup_single_snapshot(self, dataset_name: str, snapshot_name: str, parent_restic_snapshot: Optional[str]):
+        self._pre(dataset_name)
+        self._backup_single_snapshot(dataset_name, snapshot_name, parent_restic_snapshot)
+        self._post(dataset_name)
+
+    def _find_newest_snapshot_in_restic(self, snapshots_in_restic: Dict[str, str]):
+        snapshots = list(snapshots_in_restic.keys())
+        snapshots.sort()
+        return snapshots[-1]
+
+    def _find_next_snapshot(self, dataset_name: str, snapshots: List[str], newest_snapshot_in_restic: str) -> Optional[str]:
+        oldest_found = None
+        for snapshot in snapshots:
+            if snapshot <= newest_snapshot_in_restic:
+                continue
+            if oldest_found is not None and snapshot > oldest_found:
+                # snapshot is newer than another we found
+                continue
+            if self._get_zfs_snapshot_size(dataset_name, snapshot) == 0:
+                print(F"Skipping snapshot {dataset_name}@{snapshot} because of zero size.")
+                continue
+            oldest_found = snapshot
+        return oldest_found
+
+    def _backup_next_snapshot_from_dataset(self, dataset_name) -> bool:
+        restic_repo, _ = self._get_repo_name_and_path(dataset_name)
+
+        snapshots = self._get_dataset_snapshots(dataset_name)
+        snapshots_in_restic = self._get_snapshots_in_restic(restic_repo)
+        parent_restic_snapshot = None
+        newest_snapshot_in_restic = "00000000"
+        if len(snapshots_in_restic) > 0:
+            newest_snapshot_in_restic = self._find_newest_snapshot_in_restic(snapshots_in_restic)
+            parent_restic_snapshot = snapshots_in_restic[newest_snapshot_in_restic]
+        snapshot = self._find_next_snapshot(dataset_name, snapshots, newest_snapshot_in_restic)
```

Comment:
Bug: When we only look for ZFS snapshots that are newer than the newest Restic snapshot, then we skip all older ZFS snapshots that are missing in Restic (e.g. cause they were deleted in Restic or we re-run the script on a backup host with a different set of snapshots).